src/stm32_generic/i2c.cpp

```diff
@@ -0,0 +1,548 @@
+  /// Cleared by software by either a read or write in the DR register or by
+  /// hardware after a start or a stop condition in transmission or when PE=0.
+  static constexpr auto byte_transfered_finish = hal::bit_mask::from<2>();
+  /// This bit is cleared by software reading SR1 register followed reading SR2,
+  /// or by hardware when PE=0. This interrupt is generated when:
+  /// 0: No end of address transmission
+  /// 1: End of address transmission
+  static constexpr auto addr = hal::bit_mask::from<1>();
+  /// Cleared by software by reading the SR1 register followed by writing the DR
+  /// register, or by hardware when peripheral_enable=0
+  /// 0: No Start condition
+  /// 1: Start condition generated.
+  static constexpr auto start = hal::bit_mask::from<0>();
+};
+struct i2c_sr2
+{
+
+  /// 0: No General Call
+  /// 1: General Call Address received when ENGC=1
+  /// Cleared by hardware after a Stop condition or repeated Start condition, or
+  /// when PE=0
+  static constexpr auto general_call_addr = hal::bit_mask::from<4>();
+  /// 0: No communication on the bus
+  /// 1: Communication ongoing on the bus
+  /// It indicates a communication in progress on the bus. This information is
+  /// still updated when the interface is disabled (PE=0).
+  static constexpr auto trans_reciever = hal::bit_mask::from<2>();
+  /// 0: No communication on the bus
+  /// 1: Communication ongoing on the bus
+  /// It indicates a communication in progress on the bus. This information is
+  /// still updated when the interface is disabled (PE=0).
+  static constexpr auto bus_busy = hal::bit_mask::from<1>();
+  /// 0: Slave Mode
+  /// 1: Master Mode
+  /// Cleared by hardware after detecting a Stop condition on the bus or a loss
+  /// of arbitration (ARLO=1), or by hardware when PE=0.
+  static constexpr auto master = hal::bit_mask::from<0>();
+};
+struct i2c_ccr
+{
+  /// 0: Slow mode mode I2C
+  /// 1: Fast mode mode I2C
+  static constexpr auto i2c_mode_sel = hal::bit_mask::from<14>();
+  /// 0: Fm mode tlow/thigh = 2
+  /// 1: Fm mode tlow/thigh = 16/9 (see CCR)
+  static constexpr auto duty_cycle = hal::bit_mask::from<14>();
+  /// refer to CH 18.6.8 in the User manual () for equation
+  static constexpr auto ccr = hal::bit_mask::from<11, 0>();
+};
+struct i2c_filter
+{
+  /// 0 : analog filter enable
+  /// 1 : analog filter disable
+  static constexpr auto analog = hal::bit_mask::from<4>();
+
+  /// Digital noise filter enabled and filtering capability up to n * TPCLK1
+  static constexpr auto digital_filter = hal::bit_mask::from<3, 0>();
+};
+struct i2c_reg_t
+{
+  hal::u32 volatile cr1;
+  hal::u32 volatile cr2;
+  hal::u32 volatile oar1;
+  hal::u32 volatile oar2;
+  hal::u32 volatile data_reg;
+  hal::u32 volatile sr1;
+  hal::u32 volatile sr2;
+  hal::u32 volatile ccr;
+  hal::u32 volatile trise;
+  hal::u32 volatile filter;
+};
+static inline i2c_reg_t* get_i2c_reg(void* p_i2c)
+{
+  return reinterpret_cast<i2c_reg_t*>(p_i2c);
+}
```

Comment:
```suggestion
inline i2c_reg_t* get_i2c_reg(void* p_i2c)
{
  return reinterpret_cast<i2c_reg_t*>(p_i2c);
}
```

use anon namespaces over static APIs.